mithril-client-wasm/src/certificate_verification_cache.rs

```diff
@@ -0,0 +1,441 @@
+        format!("{}{}", self.cache_key_prefix, certificate_hash)
+    }
+}
+
+fn open_local_storage() -> MithrilResult<Option<Storage>> {
+    let window = window()
+        .with_context(|| "No window object")?
+        .local_storage()
+        .map_err(|err| anyhow!("Error accessing local storage: {err:?}"))?;
+    Ok(window)
+}
+
+#[cfg_attr(target_family = "wasm", async_trait(?Send))]
+#[cfg_attr(not(target_family = "wasm"), async_trait)]
+impl CertificateVerifierCache for LocalStorageCertificateVerifierCache {
+    async fn store_validated_certificate(
+        &self,
+        certificate_hash: &CertificateHash,
+        previous_certificate_hash: &PreviousCertificateHash,
+    ) -> MithrilResult<()> {
+        self.push(
+            certificate_hash,
+            previous_certificate_hash,
+            Utc::now().add(self.expiration_delay),
+        )?;
+        Ok(())
+    }
+
+    async fn get_previous_hash(
+        &self,
+        certificate_hash: &CertificateHash,
+    ) -> MithrilResult<Option<String>> {
+        match open_local_storage()? {
+            None => return Ok(None),
+            Some(storage) => {
+                let key = self.cache_key(certificate_hash);
+                match storage.get_item(&key).map_err(|err| {
+                    anyhow!("Error accessing key `{key}` from local storage: {err:?}")
+                })? {
+                    Some(value) => {
+                        let cached = Self::parse_cached_certificate(value)?;
+                        if Utc::now() >= cached.expire_at {
+                            Ok(None)
+                        } else {
+                            Ok(Some(cached.previous_hash))
+                        }
+                    }
+                    None => Ok(None),
+                }
+            }
+        }
+    }
+
+    async fn reset(&self) -> MithrilResult<()> {
+        if let Some(storage) = open_local_storage()? {
+            let len = storage
+                .length()
+                .map_err(|err| anyhow!("Error accessing local storage length: {err:?}"))?;
+            let mut key_to_remove = vec![];
+
+            for i in 0..len {
+                match storage.key(i).map_err(|err| {
+                    anyhow!("Error accessing key index `{i}` from local storage: {err:?}")
+                })? {
+                    Some(key) if key.starts_with(&self.cache_key_prefix) => key_to_remove.push(key),
+                    _ => continue,
+                }
+            }
+
+            for key in key_to_remove {
+                storage.remove_item(&key).map_err(|err| {
+                    anyhow!("Error removing key `{key}` from local storage: {err:?}")
+                })?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod test_tools {
+    use std::collections::HashMap;
+
+    use super::*;
+
+    /// `Test only` Return the raw content of the local storage
```

Comment:
Are `Test only` comments here and below mandatory since there's a `#[cfg(test)]` on the module?
```suggestion
    /// Return the raw content of the local storage
```

Reply:
Even if you need to be in a `test` context, those methods are available at the crate level (as other tests can, and do, use some of them) so I put those as a reminder but I don't mind removing them.

Reply:
I'm fine with that. Let's keep it like this 👍.